src/mlpack/methods/decision_tree/best_binary_categorical_split_impl.hpp

```diff
@@ -0,0 +1,414 @@
+     */
+    double leftGain, rightGain, gain;
+
+    // The gain for children Tₗ and Tᵣ. 
+    leftGain = FitnessFunction::template EvaluatePtr<false>(
+        classCounts.colptr(LEFT), numClasses, totalLeft
+    );
+    rightGain = FitnessFunction::template EvaluatePtr<false>(
+        classCounts.colptr(RIGHT), numClasses, totalRight
+    );
+    // The gain for this split. 
+    gain = double(totalLeft)*leftGain + double(totalRight)*rightGain;
+
+    // This is the best split found thus far. 
+    if (gain > bestFoundGain)
+    {
+      bestFoundGain = gain;
+      splitInfo.set_size(numCategories);
+      for (size_t i = 0; i < numCategories; ++i)
+        splitInfo[i] = categories[i];
+      return true;
+    }
+    return false;
+  }
+  bool improved = false;
+  uvec samples = find(categorySamples.col(k));
+
+  /**
+   * Compute the gain with category Cₖ ∈ Tₗ
+   */
+  categories[k] = LEFT;
+  for (auto i : samples)
+  {
+    ++classCounts(labels[i], LEFT);
+  }
+  totalLeft += samples.n_elem;
+
+  improved = PartitionSplit(
+      data, labels, numCategories, numClasses, bestFoundGain, categorySamples, 
+      categories, splitInfo, classCounts, totalLeft, totalRight, k+1
+  );
+
+  /** 
+   * Compute the gain with category Cₖ ∈ Tᵣ 
+   */
+  categories[k] = RIGHT;
+  for (auto i : samples)
+  {
+    --classCounts(labels[i], LEFT);
+    ++classCounts(labels[i], RIGHT);
+  }
+  totalLeft -= samples.n_elem;
+  totalRight += samples.n_elem;
+
+  improved |= PartitionSplit(
+      data, labels, numCategories, numClasses, bestFoundGain, categorySamples, 
+      categories, splitInfo, classCounts, totalLeft, totalRight, k+1
+  );
+
+  /* Unassign Cₖ and return whether a better split was found. */
+  for (auto i : samples)
+  {
+    --classCounts(labels[i], RIGHT);
+  }
+  totalRight -= samples.n_elem;
+  return improved;
+}
+
+
+// Overload used with weights.
+template<typename FitnessFunction>
+template<typename VecType, typename LabelsType, typename WeightVecType>
+bool BestBinaryCategoricalSplit<FitnessFunction>::PartitionSplit(
+      const VecType& data,
+      const LabelsType& labels,
+      const size_t numCategories,
+      const size_t numClasses,
+      const WeightVecType& weights,
+      const double totalWeight,
+      double& bestFoundGain,
+      SpMat<short>& categorySamples,
+      uvec& categories,
+      vec& splitInfo,
+      mat& classWeightSums,
+      double totalLeftWeight,
+      double totalRightWeight,
+      size_t k)
+{
+  /* Base case -- We have already found an optimal split. */
+  if (bestFoundGain >= 0.0)
+    return false;
+
+  /**
+   * Base case -- Compute the gain for the current partition.
+   */
+  if (k == numCategories)
+  {
+    double leftGain, rightGain, gain;
+
+    // The gain for children Tₗ and Tᵣ. 
+    leftGain = FitnessFunction::template EvaluatePtr<true>(
+        classWeightSums.colptr(LEFT), numClasses, totalLeftWeight
+    );
```

Comment:
If you want to simplify the code, it would be possible to merge this with the previous `PartitionSplit` version, include an extra `bool UseWeights` template parameter, and then do something like `(UseWeights ? weights[i] : 1.0)` as needed through the function.  Since `UseWeights` is known at compile time, it should compile to the same code overall.  There may be a couple more small details to handle, but I think it could work and simplify/shorten the code.

Reply:
That's actually precisely what I did initially. However, from my perspective it seems that this variant is simpler to understand, partially because the weighted algorithm requires bookkeeping for several other values, and otherwise the function signature (which is already quite big) becomes much too big for my taste. For me, the function, which is a non-trivial recursion, was much simpler to read this way, but obviously that's very subjective. If you guys would like, I could definitely implement it the other way, but my vote is to have the two overloads.